source/slang/slang-check-expr.cpp

```diff
@@ -1906,6 +1906,20 @@ namespace Slang
     {
         auto baseExpr = checkBaseForMemberExpr(subscriptExpr->baseExpression);
 
+        // If the base expression is a type, it means that this is an array declaration,
+        // then we should disable short-circuit in case there is logical expression in
+        // the subscript
+        auto baseType = baseExpr->type.Ptr();
+        auto baseTypeType = as<TypeType>(baseType);
```

Comment:
Reading this function again, now I understand why you want to avoid the extra recursion. Another simpler way to do this is something like:
```
auto subVisitor = baseTypeType?SemanticsVisitor(disableShortCircuitLogicalExpr()) : *this;

for (auto& arg : subscriptExpr->indexExprs)
   arg = subVisitor.CheckTerm(arg);
```